adapters/repos/db/vector/compressionhelpers/compression.go

```diff
@@ -0,0 +1,316 @@
+	compressor.cache.Prefetch(id)
+}
+
+func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedVectors(x, y []T) (float32, error) {
+	return compressor.quantizer.DistanceBetweenCompressedVectors(x, y)
+}
+
+func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedAndUncompressedVectors(x []T, y []float32) (float32, error) {
+	return compressor.quantizer.DistanceBetweenCompressedAndUncompressedVectors(y, x)
+}
+
+func (compressor *quantizedVectorsCompressor[T]) compressedVectorFromID(ctx context.Context, id uint64) ([]T, error) {
+	compressedVector, err := compressor.cache.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if len(compressedVector) == 0 {
+		return nil, fmt.Errorf("got a nil or zero-length vector at docID %d", id)
+	}
+	return compressedVector, nil
+}
+
+func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedVectorsFromIDs(ctx context.Context, id1, id2 uint64) (float32, error) {
+	compressedVector1, err := compressor.compressedVectorFromID(ctx, id1)
+	if err != nil {
+		return 0, err
+	}
+
+	compressedVector2, err := compressor.compressedVectorFromID(ctx, id2)
+	if err != nil {
+		return 0, err
+	}
+
+	dist, err := compressor.DistanceBetweenCompressedVectors(compressedVector1, compressedVector2)
+	return dist, err
+}
+
+func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedAndUncompressedVectorsFromID(ctx context.Context, id uint64, vector []float32) (float32, error) {
+	compressedVector, err := compressor.compressedVectorFromID(ctx, id)
+	if err != nil {
+		return 0, err
+	}
+
+	dist, err := compressor.DistanceBetweenCompressedAndUncompressedVectors(compressedVector, vector)
+	return dist, err
+}
+
+func (compressor *quantizedVectorsCompressor[T]) getCompressedVectorForID(ctx context.Context, id uint64) ([]T, error) {
+	idBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBytes, id)
+	compressedVector, err := compressor.compressedStore.Bucket(helpers.VectorsCompressedBucketLSM).Get(idBytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "Getting vector for id")
+	}
+	if len(compressedVector) == 0 {
+		return nil, storobj.NewErrNotFoundf(id, "getCompressedVectorForID")
+	}
+
+	return compressor.quantizer.FromCompressedBytes(compressedVector), nil
+}
+
+func (compressor *quantizedVectorsCompressor[T]) NewDistancer(vector []float32) (CompressorDistancer, ReturnDistancerFn) {
+	d := &quantizedCompressorDistancer[T]{
+		compressor: compressor,
+		distancer:  compressor.quantizer.NewQuantizerDistancer(vector),
+	}
+	return d, func() {
+		// ReturnDistancer method can be changed to private and removed from interface, not to be used directly
+		compressor.returnDistancer(d)
+	}
+}
+
+func (compressor *quantizedVectorsCompressor[T]) NewDistancerFromID(id uint64) (CompressorDistancer, ReturnDistancerFn) {
+	compressedVector, _ := compressor.getCompressedVectorForID(context.Background(), id)
```

Comment:
shouldn't be `compressedVectorFromID` to use cached vectors?

Reply:
Good catch, I need to see how much of performance we were loosing here...